sdk/src/main/scala/com/horizen/account/state/WithdrawalMsgProcessor.scala

```diff
@@ -0,0 +1,195 @@
+    val key = getWithdrawalEpochCounterKey(epochNum)
+    val wrCounterInBytes = view.getAccountStorageBytes(fakeSmartContractAddress.address(), key).get
+    val numOfWithdrawalReqs = if (wrCounterInBytes.length < Ints.BYTES) {
+      0
+    }
+    else {
+      Ints.fromByteArray(wrCounterInBytes)
+    }
+    numOfWithdrawalReqs
+  }
+
+  override private[horizen] def getListOfWithdrawalReqRecords(epochNum: Int, view: AccountStateView): Seq[WithdrawalRequest] = {
+    val numOfWithdrawalReqs: Int = getWithdrawalEpochCounter(view, epochNum)
+
+    val listOfWithdrawalReqs = (1 to numOfWithdrawalReqs).map(index => {
+      val currentKey = getWithdrawalRequestsKey(epochNum, index)
+      WithdrawalRequestSerializer.parseBytes(view.getAccountStorageBytes(fakeSmartContractAddress.address(), currentKey).get)
+    })
+    listOfWithdrawalReqs
+  }
+
+  protected def getListOfWithdrawalReqRecords(msg: Message, view: AccountStateView): Try[ExecutionResult] = {
+    Try {
+      require(msg.getData.length == OP_CODE_LENGTH + Ints.BYTES, s"Wrong data length ${msg.getData.length}")
+      val epochNum = BytesUtils.getInt(msg.getData, OP_CODE_LENGTH)
+
+      val listOfWithdrawalReqs = getListOfWithdrawalReqRecords(epochNum, view)
+
+      val withdrawalRequestSerializer = new ListSerializer[WithdrawalRequest](WithdrawalRequestSerializer)
+      val list = withdrawalRequestSerializer.toBytes(listOfWithdrawalReqs.asJava)
+      new ExecutionSucceeded(gasSpentForGetListOfWithdrawalReqsCmd, list)
+    }
+
+  }
+
+
+
+  private[horizen] def checkWithdrawalRequestValidity(msg: Message, view: AccountStateView): Unit = {
+    val withdrawalAmount = msg.getValue
+
+    if (msg.getData.length != OP_CODE_LENGTH + MCPublicKeyHashProposition.KEY_LENGTH) {
+      log.error(s"Wrong message data field length: ${msg.getData.length}")
+      throw new IllegalArgumentException("Wrong message data field length")
+    }
+    else if (!ZenWeiConverter.isValidZenAmount(withdrawalAmount)) {
+      log.error(s"Withdrawal amount is not a valid Zen amount: $withdrawalAmount")
+      throw new IllegalArgumentException("Withdrawal amount is not a valid Zen amount")
+    }
+    else if (withdrawalAmount.compareTo(DUST_THRESHOLD_IN_WEI) < 0) {
+      log.error(s"Withdrawal amount is under the dust threshold: $withdrawalAmount")
+      throw new IllegalArgumentException("Withdrawal amount is under the dust threshold")
+    }
+    else {
+      val balance = view.getBalance(msg.getFrom.address()).get
+      if (balance.compareTo(withdrawalAmount) < 0) {
+        log.error(s"Insufficient balance amount: balance: $balance, requested withdrawal amount: $withdrawalAmount")
+        throw new IllegalArgumentException("Insufficient balance amount")
+      }
+    }
+
+  }
+
+  protected def addWithdrawalRequest(msg: Message, view: AccountStateView): Try[ExecutionResult] = {
+    Try {
+      checkWithdrawalRequestValidity(msg, view)
```

Comment:
why not to catch the result here and return `ExecutionFailed`.
Currently we catch some cases in this methods, some in the caller.